[Core] Remove 6 month password expiry

johnsaigle opened the pull request:
## Summary
This PR removes the Password_expiry date column and adds a PasswordExpired boolean column.

Passwords will no longer be required to reset after 6 months as this is not actually a security benefit (see discussion in linked issue) and because most projects were subverting this by setting passwords to have an infinite expiry time anyway.

Calling code and relevant tables are updated.

## Testing instructions (if applicable)

Follow the Test plan at modules/login/test/Reset_Password_Test_Plan.md

## Links to related tickets (GitHub, Redmine, ...)

Resolves #4458
Resolves #4208

johnsaigle commented:
MySQL doesn't seem to start properly but I'm not sure why. I think I've updated Raisinbread properly.

ridz1208 commented:
@johnsaigle di you get it to work ?
if not I can take a look

johnsaigle commented:
Nope I'm still having trouble with it. I've tried re-initializing my local docker file and also refreshing raisinbread but I'm still getting the issue.

PapillonMcGill commented on SQL/New_patches/2019-12-17-RemovePasswordExpiry.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE users DROP COLUMN Password_expiry;
+ALTER TABLE users ADD COLUMN Password_expired TINYINT(1) NOT NULL DEFAULT 0;
```
```suggestion
ALTER TABLE users ADD COLUMN Password_expired TINYINT(1) NOT NULL DEFAULT 0 AFTER Password_hash;
```
to keep the column same order.

PapillonMcGill commented on modules/user_accounts/php/edit_user.class.inc:
```diff
@@ -254,8 +254,8 @@ class Edit_User extends \NDB_Form
-            $values['Password_hash']   = \Utility::randomString();
-            $values['Password_expiry'] = '1990-04-01';
+            $values['Password_hash']    = \Utility::randomString();
+            $values['Password_expired'] = true;
```
```suggestion
            $values['Password_expired'] = 1;
```
Although 1 can be compared to true,  tinyint is a number type. 
MySQL 5.7 accept true and false as 1 and 0 for tinyint but it might not be always the case. Safer to use 0 and 1.

driusan replied:
True and false are semantically correct. MySQL converts "BOOLEAN" datatypes to `TINYINT(1)`

PapillonMcGill commented on modules/user_accounts/php/edit_user.class.inc:
```diff
@@ -491,17 +491,11 @@ class Edit_User extends \NDB_Form
                     htmlspecialchars_decode($values['Password_hash'])
                 ),
-                $expiry
+                $values['Password_expired'] ?? false
```
```suggestion
                $values['Password_expired'] ?? 0
```
same here and more place down

kongtiaowang commented on SQL/0000-00-00-schema.sql:
```diff
@@ -107,8 +107,8 @@ CREATE TABLE `users` (
 
-INSERT INTO `users` (ID,UserID,Real_name,First_name,Last_name,Email,Privilege,PSCPI,DBAccess,Active,Pending_approval,Password_expiry)
-VALUES (1,'admin','Admin account','Admin','account','admin@example.com',0,'N','','Y','N','2016-03-30');
+INSERT INTO `users` (ID,UserID,Real_name,First_name,Last_name,Email,Privilege,PSCPI,DBAccess,Active,Pending_approval,Password_expired);
```
Remove the semicolon at the end of this line.

johnsaigle replied:
Thanks!

johnsaigle commented on SQL/0000-00-00-schema.sql:
```diff
@@ -107,8 +107,8 @@ CREATE TABLE `users` (
 
-INSERT INTO `users` (ID,UserID,Real_name,First_name,Last_name,Email,Privilege,PSCPI,DBAccess,Active,Pending_approval,Password_expiry)
-VALUES (1,'admin','Admin account','Admin','account','admin@example.com',0,'N','','Y','N','2016-03-30');
+INSERT INTO `users` (ID,UserID,Real_name,First_name,Last_name,Email,Privilege,PSCPI,DBAccess,Active,Pending_approval,Password_expired);
```
```suggestion
INSERT INTO `users` (ID,UserID,Real_name,First_name,Last_name,Email,Privilege,PSCPI,DBAccess,Active,Pending_approval,Password_expired)
```

johnsaigle replied:
```suggestion
INSERT INTO `users` (ID,UserID,Real_name,First_name,Last_name,Email,Privilege,PSCPI,DBAccess,Active,Pending_approval,Password_expired)
```

maltheism commented on modules/login/php/passwordreset.class.inc:
```diff
@@ -72,8 +72,7 @@ class PasswordReset extends \NDB_Form
                 $user->updatePassword(
-                    new \Password($password),
-                    new \DateTime('1999-01-01')
+                    new \Password($password)
```
Hi @johnsaigle the updatePassword function in php/libraries/User.class.inc has the $expiry being option but it seems to get set 6 months from now if not supplied. Is this correct with what you're trying to do with this PR?

johnsaigle replied:
This PR also contains changes to the User class where $expiry is removed in favour of $expired. This new variable represents a bool instead of a date.

maltheism commented:
@johnsaigle I'm manually testing the reset password on the login module. I'm getting this output in my error_log:
```
[Mon Jan 06 12:50:53.105137 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP Warning:  PDOStatement::execute(): SQLSTATE[22007]: Invalid datetime format: 1366 Incorrect integer value: '' for column `major`.`users`.`Password_expired` at row 1 in /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/Database.class.inc on line 585, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105509 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP Stack trace:, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105544 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   1. {main}() /Users/alizee/Development/GitHub/McGill/Loris/htdocs/index.php:0, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105572 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   2. LORIS\\Middleware\\ContentLength->process($request = *uninitialized*, $handler = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/htdocs/index.php:47, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105579 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   3. LORIS\\Middleware\\ResponseGenerator->process($request = *uninitialized*, $handler = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/src/Middleware/ContentLength.php:52, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105585 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   4. LORIS\\Router\\BaseRouter->handle($request = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/src/Middleware/ResponseGenerator.php:50, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105592 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   5. LORIS\\Router\\ModuleRouter->handle($request = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/src/Router/BaseRouter.php:100, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105619 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   6. LORIS\\login\\Module->handle($request = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/src/Router/ModuleRouter.php:99, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105625 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   7. LORIS\\login\\PasswordReset->process($request = *uninitialized*, $handler = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/Module.class.inc:342, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105631 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   8. LORIS\\Middleware\\PageDecorationMiddleware->process($request = *uninitialized*, $handler = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/NDB_Page.class.inc:695, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105655 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   9. LORIS\\Middleware\\AnonymousPageDecorationMiddleware->process($request = *uninitialized*, $handler = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/src/Middleware/PageDecorationMiddleware.php:42, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105661 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP  10. LORIS\\login\\PasswordReset->handle($request = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/src/Middleware/AnonymousPageDecorationMiddleware.php:59, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105666 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP  11. LORIS\\login\\PasswordReset->save() /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/NDB_Form.class.inc:192, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105672 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP  12. LorisForm->process($func = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/NDB_Form.class.inc:133, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105678 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP  13. LORIS\\login\\PasswordReset->_save($values = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/LorisForm.class.inc:1439, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105698 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP  14. LORIS\\login\\PasswordReset->_process($values = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/NDB_Form.class.inc:162, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105727 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP  15. User->updatePassword($password = *uninitialized*, $expired = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/modules/login/php/passwordreset.class.inc:75, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105733 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP  16. User->update($set = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/User.class.inc:503, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105739 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP  17. Database->update($table = *uninitialized*, $set = *uninitialized*, $i_where = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/User.class.inc:193, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105745 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP  18. Database->_realupdate($table = *uninitialized*, $set = *uninitialized*, $i_where = *uninitialized*, $autoescape = *uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/Database.class.inc:498, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.105752 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP  19. PDOStatement->execute(*uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/Database.class.inc:585, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:53.107323 2020] [php7:notice] [pid 5156] [client ::1:52583] Could not execute UPDATE users SET `Password_hash`=:set_Password_hash,`Password_expired`=:set_Password_expired WHERE `UserID`=:where_UserID. Stack trace#0 /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/Database.class.inc(498): Database->_realupdate('users', Array, Array, true)\n#1 /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/User.class.inc(193): Database->update('users', Array, Array)\n#2 /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/User.class.inc(503): User->update(Array)\n#3 /Users/alizee/Development/GitHub/McGill/Loris/modules/login/php/passwordreset.class.inc(75): User->updatePassword(Object(Password))\n#4 /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/NDB_Form.class.inc(162): LORIS\\login\\PasswordReset->_process(Array)\n#5 /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/LorisForm.class.inc(1439): NDB_Form->_save(Array)\n#6 /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/NDB_Form.class.inc(133): LorisForm->process(Array)\n#7 /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/NDB_Form.class.inc(192): NDB_Form->save()\n#8 /Users/alizee/Development/GitHub/McGill/Loris/src/Middleware/AnonymousPageDecorationMiddleware.php(59): NDB_Form->handle(Object(Zend\\Diactoros\\ServerRequest))\n#9 /Users/alizee/Development/GitHub/McGill/Loris/src/Middleware/PageDecorationMiddleware.php(42): LORIS\\Middleware\\AnonymousPageDecorationMiddleware->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\login\\PasswordReset))\n#10 /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/NDB_Page.class.inc(695): LORIS\\Middleware\\PageDecorationMiddleware->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\login\\PasswordReset))\n#11 /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/Module.class.inc(342): NDB_Page->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\login\\PasswordReset))\n#12 /Users/alizee/Development/GitHub/McGill/Loris/src/Router/ModuleRouter.php(99): Module->handle(Object(Zend\\Diactoros\\ServerRequest))\n#13 /Users/alizee/Development/GitHub/McGill/Loris/src/Router/BaseRouter.php(100): LORIS\\Router\\ModuleRouter->handle(Object(Zend\\Diactoros\\ServerRequest))\n#14 /Users/alizee/Development/GitHub/McGill/Loris/src/Middleware/ResponseGenerator.php(50): LORIS\\Router\\BaseRouter->handle(Object(Zend\\Diactoros\\ServerRequest))\n#15 /Users/alizee/Development/GitHub/McGill/Loris/src/Middleware/ContentLength.php(52): LORIS\\Middleware\\ResponseGenerator->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\Router\\BaseRouter))\n#16 /Users/alizee/Development/GitHub/McGill/Loris/htdocs/index.php(47): LORIS\\Middleware\\ContentLength->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\Router\\BaseRouter))\n#17 {main}, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:55.183701 2020] [php7:notice] [pid 5156] [client ::1:52583] Unhandled exception in reset_password: `Update statement did not execute successfully.`, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:55.209731 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP Warning:  Cannot modify header information - headers already sent by (output started at /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/Database.class.inc:585) in /Users/alizee/Development/GitHub/McGill/Loris/htdocs/index.php on line 53, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:55.209798 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP Stack trace:, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:55.209816 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   1. {main}() /Users/alizee/Development/GitHub/McGill/Loris/htdocs/index.php:0, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:55.209828 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   2. header(*uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/htdocs/index.php:53, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:55.209992 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP Warning:  Cannot modify header information - headers already sent by (output started at /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/Database.class.inc:585) in /Users/alizee/Development/GitHub/McGill/Loris/htdocs/index.php on line 59, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:55.210029 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP Stack trace:, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:55.210040 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   1. {main}() /Users/alizee/Development/GitHub/McGill/Loris/htdocs/index.php:0, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:55.210050 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   2. header(*uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/htdocs/index.php:59, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:55.210154 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP Warning:  Cannot modify header information - headers already sent by (output started at /Users/alizee/Development/GitHub/McGill/Loris/php/libraries/Database.class.inc:585) in /Users/alizee/Development/GitHub/McGill/Loris/htdocs/index.php on line 59, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:55.210188 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP Stack trace:, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:55.210203 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   1. {main}() /Users/alizee/Development/GitHub/McGill/Loris/htdocs/index.php:0, referer: http://localhost/login/password-reset/
[Mon Jan 06 12:50:55.210221 2020] [php7:notice] [pid 5156] [client ::1:52583] PHP   2. header(*uninitialized*) /Users/alizee/Development/GitHub/McGill/Loris/htdocs/index.php:59, referer: http://localhost/login/password-reset/

```
Inside the php/libraries/User.class.inc on line 500 is when the exception is triggered.

PapillonMcGill commented:
@maltheism Have you updated your DB schema?

maltheism commented:
@PapillonMcGill I started fresh and ran the rasinbread database setup script.
```
cat SQL/0000-00-00-schema.sql \
    SQL/0000-00-01-Permission.sql \
    SQL/0000-00-02-Menus.sql \
    SQL/0000-00-03-ConfigTables.sql \
    SQL/0000-00-04-Help.sql \
    SQL/0000-00-05-ElectrophysiologyTables.sql \
    raisinbread/instruments/instrument_sql/aosi.sql \
    raisinbread/instruments/instrument_sql/bmi.sql \
    raisinbread/instruments/instrument_sql/medical_history.sql \
    raisinbread/instruments/instrument_sql/mri_parameter_form.sql \
    raisinbread/instruments/instrument_sql/radiology_review.sql \
    test/test_instrument/testtest.sql \
    raisinbread/RB_files/*.sql | mysql -u root -p -h localhost major
```

PapillonMcGill commented:
@maltheism this patch has an update to the schema that you have to manually apply.

maltheism commented:
@PapillonMcGill I'm a little confused because shouldn't the patch be executed when I run the above command on an empty database?

PapillonMcGill commented:
Yes, if you ran the raisin_bread update script on the 0000-00-00-schema.sql file modified in this PR. But the error point to an exception while updating the data in the users table. So I though that the users table do not contain the latest change.

johnsaigle commented:
So is the patch fine then?

PapillonMcGill commented:
@maltheism Could you retest with the proper SQL patch applied? See if there is still some error?

maltheism commented:
@PapillonMcGill @johnsaigle I'm trying to complete the installdb.php script because I tested it successfully on the master branch this morning. I'm getting this output when testing the installdb.php on this PR with an empty database:
```
[Thu Jan 09 12:31:43.808024 2020] [php7:notice] [pid 30541] [client 192.168.122.1:47646] PHP Warning:  PDOStatement::execute(): SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual$
[Thu Jan 09 12:31:43.808162 2020] [php7:notice] [pid 30541] [client 192.168.122.1:47646] PHP Stack trace:, referer: https://awickenheiser-dev.loris.ca/installdb.php
[Thu Jan 09 12:31:43.808191 2020] [php7:notice] [pid 30541] [client 192.168.122.1:47646] PHP   1. {main}() /var/www/Loris/htdocs/installdb.php:0, referer: https://awickenheiser-dev.loris.ca/installdb.php
[Thu Jan 09 12:31:43.808212 2020] [php7:notice] [pid 30541] [client 192.168.122.1:47646] PHP   2. LORIS\\Installer\\Installer->resetFrontEndAdmin() /var/www/Loris/htdocs/installdb.php:162, referer: https://awickenheiser-dev.loris.ca/insta$
[Thu Jan 09 12:31:43.808230 2020] [php7:notice] [pid 30541] [client 192.168.122.1:47646] PHP   3. PDOStatement->execute() /var/www/Loris/php/installer/Installer.class.inc:553, referer: https://awickenheiser-dev.loris.ca/installdb.php
```

PapillonMcGill commented:
@maltheism When you say empty DB, did you ran the SQL/0000-00-0w-xyz.sql scripts before?
Which 0000-00-02-script did you had in your SQL/ directory? (0000-00-02-Menu.sql or 0000-00-02-Modules.sql
and mostly, have you provided valid front-end user and password?
The error appended when the script tried to change the admin user (1) userID/password and set it to active.

PapillonMcGill commented on SQL/New_patches/2019-12-17-RemovePasswordExpiry.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE users DROP COLUMN Password_expiry;
+ALTER TABLE users ADD COLUMN Password_expired TINYINT(1) NOT NULL DEFAULT 0 AFTER Password_hash;
```
Add the new column,
UPDATE users SET Password_expired = 0 where Password_expiry > timestamp;
UPDATE users SET Password_expired = 1 where Password_expired IS NULL;
drop old column

Setting a default value will not change existing rows, only set the default for new entries.

PapillonMcGill replied:
@johnsaigle You also have to specify 0 where still valid. Adding a new column fill it with `null` value.

johnsaigle replied:
Won't DEFAULT 0 take care of that?

PapillonMcGill replied:
Yes, sorry, I was still tinking past version or maybe a mix with Oracle (again older version).

PapillonMcGill commented:
Tried to add a new users with this PR (DB properly patched before hand)
```
[Thu Jan 16 08:35:04.000058 2020] [php7:warn] [pid 24823] [client 132.216.122.95:40782] PHP Warning:  strlen() expects parameter 1 to be string, array given in /var/www/Loris/php/libraries/LorisForm.class.inc on line 2077, referer: https://mlegault-dev.loris.ca/user_accounts/edit_user/
[Thu Jan 16 08:35:04.000337 2020] [php7:notice] [pid 24823] [client 132.216.122.95:40782] PHP Notice:  Undefined index: NA_UserID in /var/www/Loris/modules/user_accounts/php/edit_user.class.inc on line 1029, referer: https://mlegault-dev.loris.ca/user_accounts/edit_user/
[Thu Jan 16 08:35:04.648565 2020] [php7:notice] [pid 24823] [client 132.216.122.95:40782] PHP Notice:  Undefined index: Password_hash in /var/www/Loris/php/libraries/User.class.inc on line 524, referer: https://mlegault-dev.loris.ca/user_accounts/edit_user/
[Thu Jan 16 08:35:04.650140 2020] [php7:warn] [pid 24823] [client 132.216.122.95:40782] PHP Warning:  strlen() expects parameter 1 to be string, array given in /var/www/Loris/php/libraries/LorisForm.class.inc on line 2077, referer: https://mlegault-dev.loris.ca/user_accounts/edit_user/
[Thu Jan 16 08:35:04.653063 2020] [php7:notice] [pid 24823] [client 132.216.122.95:40782] PHP Notice:  Undefined index: ID in /var/www/Loris/php/libraries/User.class.inc on line 225, referer: https://mlegault-dev.loris.ca/user_accounts/edit_user/
[Thu Jan 16 08:35:05.325300 2020] [php7:warn] [pid 24823] [client 132.216.122.95:40782] PHP Warning:  PDOStatement::execute(): SQLSTATE[HY000]: General error: 1366 Incorrect integer value: '' for column 'Password_expired' at row 1 in /var/www/Loris/php/libraries/Database.class.inc on line 585, referer: https://mlegault-dev.loris.ca/user_accounts/edit_user/
[Thu Jan 16 08:35:05.325610 2020] [php7:notice] [pid 24823] [client 132.216.122.95:40782] Could not execute UPDATE users SET `Password_hash`=:set_Password_hash,`Password_expired`=:set_Password_expired WHERE `UserID`=:where_UserID. Stack trace#0 /var/www/Loris/php/libraries/Database.class.inc(498): Database->_realupdate('users', Array, Array, true)\n#1 /var/www/Loris/php/libraries/User.class.inc(193): Database->update('users', Array, Array)\n#2 /var/www/Loris/php/libraries/User.class.inc(503): User->update(Array)\n#3 /var/www/Loris/modules/user_accounts/php/edit_user.class.inc(498): User->updatePassword(Object(Password), false)\n#4 /var/www/Loris/php/libraries/NDB_Form.class.inc(162): LORIS\\user_accounts\\Edit_User->_process(Array)\n#5 /var/www/Loris/php/libraries/LorisForm.class.inc(1439): NDB_Form->_save(Array)\n#6 /var/www/Loris/php/libraries/NDB_Form.class.inc(133): LorisForm->process(Array)\n#7 /var/www/Loris/php/libraries/NDB_Form.class.inc(192): NDB_Form->save()\n#8 /var/www/Loris/modules/user_accounts/php/edit_user.class.inc(604): NDB_Form->handle(Object(Zend\\Diactoros\\ServerRequest))\n#9 /var/www/Loris/src/Middleware/UserPageDecorationMiddleware.php(189): LORIS\\user_accounts\\Edit_User->handle(Object(Zend\\Diactoros\\ServerRequest))\n#10 /var/www/Loris/src/Middleware/PageDecorationMiddleware.php(53): LORIS\\Middleware\\UserPageDecorationMiddleware->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\user_accounts\\Edit_User))\n#11 /var/www/Loris/php/libraries/NDB_Page.class.inc(695): LORIS\\Middleware\\PageDecorationMiddleware->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\user_accounts\\Edit_User))\n#12 /var/www/Loris/php/libraries/Module.class.inc(342): NDB_Page->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\user_accounts\\Edit_User))\n#13 /var/www/Loris/modules/user_accounts/php/module.class.inc(43): Module->handle(Object(Zend\\Diactoros\\ServerRequest))\n#14 /var/www/Loris/src/Middleware/ResponseGenerator.php(50): LORIS\\user_accounts\\Module->handle(Object(Zend\\Diactoros\\ServerRequest))\n#15 /var/www/Loris/src/Middleware/AuthMiddleware.php(63): LORIS\\Middleware\\ResponseGenerator->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\user_accounts\\Module))\n#16 /var/www/Loris/src/Router/ModuleRouter.php(96): LORIS\\Middleware\\AuthMiddleware->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\user_accounts\\Module))\n#17 /var/www/Loris/src/Router/BaseRouter.php(100): LORIS\\Router\\ModuleRouter->handle(Object(Zend\\Diactoros\\ServerRequest))\n#18 /var/www/Loris/src/Middleware/ResponseGenerator.php(50): LORIS\\Router\\BaseRouter->handle(Object(Zend\\Diactoros\\ServerRequest))\n#19 /var/www/Loris/src/Middleware/ContentLength.php(52): LORIS\\Middleware\\ResponseGenerator->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\Router\\BaseRouter))\n#20 /var/www/Loris/htdocs/index.php(47): LORIS\\Middleware\\ContentLength->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\Router\\BaseRouter))\n#21 {main}, referer: https://mlegault-dev.loris.ca/user_accounts/edit_user/
[Thu Jan 16 08:35:05.326751 2020] [php7:error] [pid 24823] [client 132.216.122.95:40782] PHP Fatal error:  Uncaught DatabaseException: Update statement did not execute successfully. in /var/www/Loris/php/libraries/Database.class.inc:587\nStack trace:\n#0 /var/www/Loris/php/libraries/Database.class.inc(498): Database->_realupdate('users', Array, Array, true)\n#1 /var/www/Loris/php/libraries/User.class.inc(193): Database->update('users', Array, Array)\n#2 /var/www/Loris/php/libraries/User.class.inc(503): User->update(Array)\n#3 /var/www/Loris/modules/user_accounts/php/edit_user.class.inc(498): User->updatePassword(Object(Password), false)\n#4 /var/www/Loris/php/libraries/NDB_Form.class.inc(162): LORIS\\user_accounts\\Edit_User->_process(Array)\n#5 /var/www/Loris/php/libraries/LorisForm.class.inc(1439): NDB_Form->_save(Array)\n#6 /var/www/Loris/php/libraries/NDB_Form.class.inc(133): LorisForm->process(Array)\n#7 /var/www/Loris/php/libraries/NDB_Form.class.inc(192): NDB_Form->save()\n#8 /var/www/Loris/modules/user_accounts/php/edit_user.class.inc(604): NDB_Form->handle(Object(Zend\\Diactoros\\ServerRequest))\n#9 /var/www/Loris/src/M in /var/www/Loris/php/libraries/Database.class.inc on line 587, referer: https://mlegault-dev.loris.ca/user_accounts/edit_user/
```

PapillonMcGill commented:
tried to add a new user (DB properly patched before hand)
```
[Thu Jan 16 08:35:04.000058 2020] [php7:warn] [pid 24823] [client 132.216.122.95:40782] PHP Warning:  strlen() expects parameter 1 to be string, array given in /var/www/Loris/php/libraries/LorisForm.class.inc on line 2077, referer: https://example.org/user_accounts/edit_user/
[Thu Jan 16 08:35:04.000337 2020] [php7:notice] [pid 24823] [client 132.216.122.95:40782] PHP Notice:  Undefined index: NA_UserID in /var/www/Loris/modules/user_accounts/php/edit_user.class.inc on line 1029, referer: https://example.org/user_accounts/edit_user/
[Thu Jan 16 08:35:04.648565 2020] [php7:notice] [pid 24823] [client 132.216.122.95:40782] PHP Notice:  Undefined index: Password_hash in /var/www/Loris/php/libraries/User.class.inc on line 524, referer: https://example.org/user_accounts/edit_user/
[Thu Jan 16 08:35:04.650140 2020] [php7:warn] [pid 24823] [client 132.216.122.95:40782] PHP Warning:  strlen() expects parameter 1 to be string, array given in /var/www/Loris/php/libraries/LorisForm.class.inc on line 2077, referer: https://example.org/user_accounts/edit_user/
[Thu Jan 16 08:35:04.653063 2020] [php7:notice] [pid 24823] [client 132.216.122.95:40782] PHP Notice:  Undefined index: ID in /var/www/Loris/php/libraries/User.class.inc on line 225, referer: https://example.org/user_accounts/edit_user/
[Thu Jan 16 08:35:05.325300 2020] [php7:warn] [pid 24823] [client 132.216.122.95:40782] PHP Warning:  PDOStatement::execute(): SQLSTATE[HY000]: General error: 1366 Incorrect integer value: '' for column 'Password_expired' at row 1 in /var/www/Loris/php/libraries/Database.class.inc on line 585, referer: https://example.org/user_accounts/edit_user/
[Thu Jan 16 08:35:05.325610 2020] [php7:notice] [pid 24823] [client 132.216.122.95:40782] Could not execute UPDATE users SET `Password_hash`=:set_Password_hash,`Password_expired`=:set_Password_expired WHERE `UserID`=:where_UserID. Stack trace#0 /var/www/Loris/php/libraries/Database.class.inc(498): Database->_realupdate('users', Array, Array, true)\n#1 /var/www/Loris/php/libraries/User.class.inc(193): Database->update('users', Array, Array)\n#2 /var/www/Loris/php/libraries/User.class.inc(503): User->update(Array)\n#3 /var/www/Loris/modules/user_accounts/php/edit_user.class.inc(498): User->updatePassword(Object(Password), false)\n#4 /var/www/Loris/php/libraries/NDB_Form.class.inc(162): LORIS\\user_accounts\\Edit_User->_process(Array)\n#5 /var/www/Loris/php/libraries/LorisForm.class.inc(1439): NDB_Form->_save(Array)\n#6 /var/www/Loris/php/libraries/NDB_Form.class.inc(133): LorisForm->process(Array)\n#7 /var/www/Loris/php/libraries/NDB_Form.class.inc(192): NDB_Form->save()\n#8 /var/www/Loris/modules/user_accounts/php/edit_user.class.inc(604): NDB_Form->handle(Object(Zend\\Diactoros\\ServerRequest))\n#9 /var/www/Loris/src/Middleware/UserPageDecorationMiddleware.php(189): LORIS\\user_accounts\\Edit_User->handle(Object(Zend\\Diactoros\\ServerRequest))\n#10 /var/www/Loris/src/Middleware/PageDecorationMiddleware.php(53): LORIS\\Middleware\\UserPageDecorationMiddleware->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\user_accounts\\Edit_User))\n#11 /var/www/Loris/php/libraries/NDB_Page.class.inc(695): LORIS\\Middleware\\PageDecorationMiddleware->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\user_accounts\\Edit_User))\n#12 /var/www/Loris/php/libraries/Module.class.inc(342): NDB_Page->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\user_accounts\\Edit_User))\n#13 /var/www/Loris/modules/user_accounts/php/module.class.inc(43): Module->handle(Object(Zend\\Diactoros\\ServerRequest))\n#14 /var/www/Loris/src/Middleware/ResponseGenerator.php(50): LORIS\\user_accounts\\Module->handle(Object(Zend\\Diactoros\\ServerRequest))\n#15 /var/www/Loris/src/Middleware/AuthMiddleware.php(63): LORIS\\Middleware\\ResponseGenerator->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\user_accounts\\Module))\n#16 /var/www/Loris/src/Router/ModuleRouter.php(96): LORIS\\Middleware\\AuthMiddleware->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\user_accounts\\Module))\n#17 /var/www/Loris/src/Router/BaseRouter.php(100): LORIS\\Router\\ModuleRouter->handle(Object(Zend\\Diactoros\\ServerRequest))\n#18 /var/www/Loris/src/Middleware/ResponseGenerator.php(50): LORIS\\Router\\BaseRouter->handle(Object(Zend\\Diactoros\\ServerRequest))\n#19 /var/www/Loris/src/Middleware/ContentLength.php(52): LORIS\\Middleware\\ResponseGenerator->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\Router\\BaseRouter))\n#20 /var/www/Loris/htdocs/index.php(47): LORIS\\Middleware\\ContentLength->process(Object(Zend\\Diactoros\\ServerRequest), Object(LORIS\\Router\\BaseRouter))\n#21 {main}, referer: https://example.org/user_accounts/edit_user/
[Thu Jan 16 08:35:05.326751 2020] [php7:error] [pid 24823] [client 132.216.122.95:40782] PHP Fatal error:  Uncaught DatabaseException: Update statement did not execute successfully. in /var/www/Loris/php/libraries/Database.class.inc:587\nStack trace:\n#0 /var/www/Loris/php/libraries/Database.class.inc(498): Database->_realupdate('users', Array, Array, true)\n#1 /var/www/Loris/php/libraries/User.class.inc(193): Database->update('users', Array, Array)\n#2 /var/www/Loris/php/libraries/User.class.inc(503): User->update(Array)\n#3 /var/www/Loris/modules/user_accounts/php/edit_user.class.inc(498): User->updatePassword(Object(Password), false)\n#4 /var/www/Loris/php/libraries/NDB_Form.class.inc(162): LORIS\\user_accounts\\Edit_User->_process(Array)\n#5 /var/www/Loris/php/libraries/LorisForm.class.inc(1439): NDB_Form->_save(Array)\n#6 /var/www/Loris/php/libraries/NDB_Form.class.inc(133): LorisForm->process(Array)\n#7 /var/www/Loris/php/libraries/NDB_Form.class.inc(192): NDB_Form->save()\n#8 /var/www/Loris/modules/user_accounts/php/edit_user.class.inc(604): NDB_Form->handle(Object(Zend\\Diactoros\\ServerRequest))\n#9 /var/www/Loris/src/M in /var/www/Loris/php/libraries/Database.class.inc on line 587, referer: https://example.org/user_accounts/edit_user/
```

PapillonMcGill commented:
tried to add a new user (DB properly patched before hand)
```

johnsaigle commented:
Thanks for testing. I'll try to see if I can reproduce it.

johnsaigle commented:
@PapillonMcGill I wasn't able to reproduce this. Could you provide more details about what you tried?

I did the following:
* Checked out the branch
* Sourced the patch: `mysql -A jsaigle_sandbox < SQL/New_patches/2019-12-17-RemovePasswordExpiry.sql`
* `make dev`
* Logged in as `admin`, added a new user
* User was created successfully

I also just rebased the branch so perhaps there was an issue with the PR not containing the most recent changes.

PapillonMcGill commented:
@johnsaigle Same problem

checkout you branch, latest commit
mysql  < SQL/New_patches/2019-12-17-RemovePasswordExpiry.sql
mysql  < SQL/New_patches/2019-12-05-AddModuleTable.sql   (fail without it)
make
php raisinbread_refresh.php

johnsaigle commented:
If you do raisinbread refresh after those steps then it will overwrite the patches you applied.

PapillonMcGill commented:
the RB_users.sql files that is loaded with raisinbread_refresh.php include the change to users table included in this PR. (Unless there is a problem with this version of the files).

johnsaigle commented:
I agree with you, just saying that you don't need to apply those patches if raisinbread is applied.

johnsaigle commented:
@PapillonMcGill We have an integration test for adding a new user and it doesn't seem to be throwing the error you encountered. I also am not able to reproduce this on my VM.

Can you try again?

PapillonMcGill commented:
tested again:

checkout aces/master
clear DB
run 0000-00-0x-scripts
make
-- working fine--
checkout this PR
run sql patch 2019-12-17-RemovePasswordExpiry.sql
make
try to add a new user
-- same error --

in the error log ```
 Could not execute UPDATE users SET `Password_hash`=:set_Password_hash,`Password_expired`=:set_Password_expired WHERE `UserID ...
```
Could it be that the usage of `false` instead of `0` be the cause?
I have MySQL 5.7.29.
php 7.3.14 (from phpinfo())

johnsaigle commented:
Can you copy the POST values you are submitting to create a new user?  Maybe you've found a path in the code that me and Travis aren't going down for some reason.

PapillonMcGill commented:
Array ( [UserID] => bis [Password_hash] => redacted [__Confirm] => redacted [First_name] => testing [Last_name] => dd [Degree] => [Position_title] => [Institution] => [Department] => [Address] => [City] => [State] => [Country] => [Fax] => [Email] => ouain@example.ca [__ConfirmEmail] => ouain@example.ca [CenterIDs] => Array ( [0] => 1 ) [ProjectIDs] => Array ( [0] => 1 ) [ex_1] => on [examiner_radiologist] => N [examiner_pending] => N [Active] => Y [active_from] => 2020-01-27 [active_to] => 2020-01-28 [Pending_approval] => N [permID] => Array ( [5] => on [11] => on [10] => on [13] => on ) [fire_away] => Save )

johnsaigle commented:
@PapillonMcGill I was able to reproduce the bug with those values. The `User::update()` code doesn't convert bools to equivalent MySQL values and instead caused it to be a string which caused the DB error.

It should be fixed now.

PapillonMcGill commented:
@johnsaigle While testing this, my test user could not check his `my preference` page without `user management` permission. 
Just checked with aces/master and it work fine. This might be a bug introduced by this PR.
I will recheck

johnsaigle commented:
Are you sure you're on the latest version of aces/master? This is a known bug #6040. I recently rebased this so it will also contain the bug.

PapillonMcGill commented:
@johnsaigle Yes, went back in history and I did fetch from aces/master just before the checkout. Also ran rainsinbread_refresh. and make

johnsaigle commented:
I did the same and the issue exists for me on master. I don't think it's related to this PR.

driusan commented:
What does the "Password_expired" bool mean if there's no more password expiry?

johnsaigle commented:
It's how we would trigger a password reset, e.g. when a one-time server-generated password is sent to a user when their account is created.

driusan commented:
Then we should name it something more accurate like `PasswordChangeRequired`

johnsaigle commented:
If https://github.com/aces/Loris/pull/6511 goes through, this PR needs to updated that document to change language about password expiry.

zaliqarosli commented:
hey @johnsaigle please tag me when you are ready for my review. thanks!

johnsaigle commented on raisinbread/RB_files/RB_users.sql:
```diff
@@ -1,6 +1,6 @@
 TRUNCATE TABLE `users`;
 LOCK TABLES `users` WRITE;
-INSERT INTO `users` (`ID`, `UserID`, `Password`, `Real_name`, `First_name`, `Last_name`, `Degree`, `Position_title`, `Institution`, `Department`, `Address`, `City`, `State`, `Zip_code`, `Country`, `Phone`, `Fax`, `Email`, `Privilege`, `PSCPI`, `DBAccess`, `Active`, `Password_hash`, `Password_expiry`, `Pending_approval`, `Doc_Repo_Notifications`, `language_preference`, `active_from`, `active_to`)  ...
+INSERT INTO `users` (`ID`, `UserID`, `Password`, `Real_name`, `First_name`, `Last_name`, `Degree`, `Position_title`, `Institution`, `Department`, `Address`, `City`, `State`, `Zip_code`, `Country`, `Phone`, `Fax`, `Email`, `Privilege`, `PSCPI`, `DBAccess`, `Active`, `Password_hash`, `PasswordChangeRequired`, `Pending_approval`, `Doc_Repo_Notifications`, `language_preference`, `active_from`, `activ ...
```
The difference here is the replacement of `2100-01-01` with `0`

johnsaigle commented:
@zaliqarosli This is ready now!

zaliqarosli commented on php/libraries/User.class.inc:
```diff
@@ -482,37 +482,29 @@ class User extends UserPermissions
     function updatePassword(
         \Password $password,
-        ?\DateTime $expiry = null
+        bool $expired = false
```
does this need to have ?bool specified if the $expired argument isn't provided in the function call?

johnsaigle replied:
I don't think so. The `?` signals that a variable can be null which isn't the case here. If an argument isn't provided, it'll default to `false`.

zaliqarosli commented on test/unittests/UserTest.php:
```diff
@@ -58,7 +58,7 @@ class UserTest extends TestCase
                 'Active'                 => 'Y',
                 'Password_hash'          => null,
-                'Password_expiry'        => '2020-07-16',
+                'PasswordChangeRequired'        => 0,
```
not sure if this needs phpcs to pass travis

johnsaigle replied:
Me neither!

johnsaigle commented:
Yep. I just updated the test plan because I forgot to before. Should be good now.

johnsaigle commented:
blocked by #6733

driusan commented:
@johnsaigle the PR you says is blocking this is already merged?

johnsaigle commented:
Yeah you merged it after I wrote that comment.
